### ISSUE TITLE
Purpose of FUNCTIONAL_* user string entries

### ISSUE THREAD

**Issue by adrianbroher** [freeorion/freeorion#907]
### Description

Currently we use the UserString.\* API to query some configuration values from files.  I do think this is a subpar solution for querying those configuration and was wondering what was the rationale behind this and why are those values not part of the OptionsDB.  The problems I see are:
- Tools to extract translation keys from the code  need to be more complex to exclude FUNCTIONAL_ entries.
- UserString self documentation implies that you give a key and get a localized string. Except for the FUNCTIONAL_ entries.
- StringTables now need to support (needlessly) a `#include` syntax to keep the configuration values out of the translators reach.
- Translation keys can't be chosen abitrary, but need to stay away from the `FUNCTIONAL_` namespace.

tl;dr: The current implementation intermingles to the tasks of 'give key, get translation' and 'give key, get configuration value'.

Only references I found so far are:

http://freeorion.org/forum/viewtopic.php?p=83590#p83590

**Comment by Vezzra** [comment]
Agree to everything you said. The discussion you linked to already contains proposals how to fix this. I agree with @Dilvish-fo that we should avoid code duplication by providing two separate interfaces in the backend code that do essentially the same technically ("give string key, get some string value").

So the idea to provide a generic implementation that can be used for value retrieval by key for both localized text and configuration values sounds very reasonable to me, we just need to move all those FUNCTIONAL_\* entries out of the stringtables and put them in dedicated configuration files (which are stored at a location that makes more sense than the stringtables...).

> what was the rationale behind this

AFAIK, because it has been quick and easy. I don't think that there is really any oder reason behind it.

**Comment by adrianbroher** [comment]
Thanks for the answer Vezzra, but unfortunately it doesn't answer why are those valuee not part of the OptionsDB.

> […] that do essentially the same technically ("give string key, get some string value").

But this isn't the case. One use case converts to float and one converts it to a list of string, which is something very different than a string.

**Comment by geoffthemedio** [comment]
The OptionsDB probably doesn't presently allow or at least handle correctly multi-line strings as a data type.

The stuff in
https://github.com/freeorion/freeorion/blob/master/default/customizations/common_user_customizations.txt
seems to be UI controls, though, so would make sense to put in OptionsDB. (I would avoid moving anything content-related into the OptionsDB, until there's a general game-rules-configuration mechanism in place).

**Comment by MatGB** [comment]
Ah, that's where they are now.

K, two seprate settings, the sitrep priority order list, which frankly doesn't work where it is, people constantly forget to add things to it, etc. Then there's the numbers @Dilvish-fo used for filtering redundancy in the design window which I've never really played with.

I vote moving the latter to optionsdb which is basically where they ought to be anyway, and putting the sitrep order somewhere front and centre within scripting where hopefully people will remember it exists more often—we don't want it back in the stringtables.

**Comment by Vezzra** [comment]
> doesn't answer why are those value not part of the OptionsDB.

In addition to what @geoffthemedio said: "quick and easy" not only refers to the effort regarding implementation, but also ease of use for those who want to tinker with these configuration values. OptionsDB isn't really easy to edit with a text editor (hacking around in a XML file is a bit tedious), the configuration values stored in there are actually meant to be manipulated via GUI elements (options dialog).

Just (ab)using the user string (list) functions and let the user edit some values/lists in the stringtables has been far easier and quicker to provide, and for the user to edit with a text editor, than adding/using entries in OptionsDB (not to speak of providing GUI elements to edit them...).

A classic example of "quick and dirty".

> One use case converts to float and one converts it to a list of string, which is something very different than a string.

True, but two use cases (retrieval of string value and retrieval of an entire list of strings) is something both the stringtable as well as these new configuration files need, the only difference would be that we don't need to retrieve numbers from stringtables. So that part of a generic implementation would never be used when accessing the stringtables, but that's no reason not to use a generic implementation for both interfaces, isn't it?

**Comment by dbenage-cx** [comment]
> ease of use for those who want to tinker with these configuration values

If editing an XML file is too taxing for the average user, then I'd argue to change `config.xml` to another format.  I don't feel XML is a high barrier here tough.  As noted, any options popular/important enough tend to receive a GUI counterpart.  Providing that GUI element for a `UserString` or `ContentString` configuration variable would need further support.

> Just (ab)using the user string (list) functions and let the user edit some values/lists...

I suspect few have ever modified many of these, namely because noone has posted an issue for [topping-up factor using the value from frontload factor](https://github.com/freeorion/freeorion/blob/master/Empire/Empire.cpp#L56) or [the part redundancy skip list appears to not be used/is hard-coded](https://github.com/freeorion/freeorion/blob/master/UI/DesignWnd.cpp#L590).

For the config additons, I've had:
- [part redundancy filter->config and tag](https://github.com/freeorion/freeorion/compare/master...dbenage-cx:functional-to-config_design_filter)
- [production queue factors->config](https://github.com/freeorion/freeorion/compare/master...dbenage-cx:functional-to-config_production_factors)
- [shipyard list->tag](https://github.com/freeorion/freeorion/compare/master...dbenage-cx:functional-to-tag_shipyard_list)

I've held off on posting PRs, as there seems to still be contention on where the values should go.
If these are not the direction to take, I'll post new issues for those mentioned above.

**Comment by adrianbroher** [comment]
> I've held off on posting PRs, as there seems to still be contention on where the values should go.

Well done @dbenage-cx , I like the solutions you propose here (except the usual nagging like "why 'BUILDING_LIST_SHIPYARD' and not just 'SHIPYARD'?") because they reuse already existing concepts like tags.

> If these are not the direction to take, I'll post new issues for those mentioned above.

I think you picked some the config values where the future placement is quite easy to agree on.

**Comment by MatGB** [comment]
The part redundancy stuff was added by Dilvish because he uses it himself and we thought having it in trunk would be good for edge case users, I didn't use it often and now we've got refinements the basic need for it has massively reduced (the main complaint was "I've researched Plasma 1 but it doesn't show up" the reason being Laser3 was clearly the better part, that's no longer the case).

Consequently, I never noticed if it worked properly or not, it's not a balance thing and I don't personally care for the feature (it's a good feature to have, just not one I need).

**Comment by Vezzra** [comment]
> I don't feel XML is a high barrier here tough.

Of course it's not insurmountable, but certainly not as readable and easy to edit as a plain list of strings.

> As noted, any options popular/important enough tend to receive a GUI counterpart.

Exactly, and those options can (and should) end up in config.xml.

> Providing that GUI element for a UserString or ContentString configuration variable would need further support.

That would be pointless, because I think that's just for configuration variables that are supposed for devs to easily tinker with some parameters, or for stuff that eventually should be moved to config.xml and only ended up in an extra config file because someone didn't want to go through the all the trouble necessary to add something to config.xml.

Most obvious reason IMO would be if implementing a feature that's still subject to change, so before you waste time and effort providing extensions to the options dialog, you use the quick and dirty method, and only move the stuff to config.xml once everything is (sufficiently) set in stone.

The problem with this is that often enough unfinished stuff/stop gap solutions remain for a very long time before getting thrown out or completed, and so do related stop-gap config solutions...
